core/ledger/kvledger/hashcheck_pvtdata.go

```diff
@@ -173,3 +193,100 @@ func findInvalidNsPvtData(nsRwset *rwset.NsPvtReadWriteSet, txRWSet *rwsetutil.T
 	}
 	return invalidPvtData, invalidNsColl
 }
+
+func verifyHashesAndTrimWriteSetViaBootKVHashes(reconciledPvtdata *ledger.ReconciledPvtdata, pvtdataStore *pvtdatastorage.Store) (
+	[]*ledger.TxPvtData, []*ledger.PvtdataHashMismatch, error,
+) {
+	var validPvtData []*ledger.TxPvtData
+	var invalidPvtData []*ledger.PvtdataHashMismatch
+
+	blkNum := reconciledPvtdata.BlockNum
+
+	for txNum, txData := range reconciledPvtdata.WriteSets { // Tx loop
+		reconTx, err := rwsetutil.TxPvtRwSetFromProtoMsg(txData.WriteSet)
+		if err != nil {
+			continue
+		}
+
+		trimmedTx := &rwsetutil.TxPvtRwSet{}
+
+		for _, reconNS := range reconTx.NsPvtRwSet { // Ns Loop
+			trimmedNs := &rwsetutil.NsPvtRwSet{
+				NameSpace: reconNS.NameSpace,
+			}
+
+			for _, reconColl := range reconNS.CollPvtRwSets { // coll loop
+				if reconColl.KvRwSet == nil || len(reconColl.KvRwSet.Writes) == 0 {
+					continue
+				}
+
+				expectedKVHashes, err := pvtdataStore.FetchBootKVHashes(blkNum, txNum, reconNS.NameSpace, reconColl.CollectionName)
+				if err != nil {
+					return nil, nil, err
+				}
+				if len(expectedKVHashes) == 0 {
+					continue
+				}
+
+				filteredInKVs := []*kvrwset.KVWrite{}
+				hashMismatch := &ledger.BootKVsMismatch{}
+
+				for _, reconKV := range reconColl.KvRwSet.Writes {
+					reconKeyHash := util.ComputeSHA256([]byte(reconKV.Key))
+					reconValHash := util.ComputeSHA256(reconKV.Value)
+
+					expectedValHash, ok := expectedKVHashes[string(reconKeyHash)]
+					if ok {
+						delete(expectedKVHashes, string(reconKeyHash))
+						if bytes.Equal(expectedValHash, reconValHash) {
+							filteredInKVs = append(filteredInKVs, reconKV)
+						} else {
+							hashMismatch.NumMismatches++
+						}
+					}
+				}
+				hashMismatch.NumKeysNotSupplied = len(expectedKVHashes)
```

Comment:
Using a separate variable to count this would add clarity.

Reply:
Sure